src/NServiceBus.Metrics.ServiceControl/ReportingFeature.cs

```diff
@@ -232,15 +232,18 @@ public ServiceControlRawDataReporting(IBuilder builder, ReportingOptions options
 
             protected override Task OnStart(IMessageSession session)
             {
-                foreach (var metric in metrics)
+                options.OnCreateReporters(() =>
```

Comment:
I might be misreading this, but if the order here is

1. `MetricOptions.OnStart`
2. `ReportingFeature.OnStart`

Then the reporters would never be created.

If `Metrics.OnStart` is called first, it will invoke the `ReportingOptions.CreateReporters` method before the action has been set via the `OnCreateReporters` method. This means that `CreateReporters()` will be invoked while `createMetricReporters` is null, so no reporters will be created.

Reply:
First, my assumption is that `metricsOptions.SendMetricDataToServiceControl(…)` is going to be called first, as this is a config method before the endpoint is initialized. This means that the lambda to execute `reporting.CreateReporters()` will definitely be stored. 

The question then, is which happens first? The FeatureStartupTask `ServiceControlRawDataReporting.OnStart()`? Or does Metrics execute the callback that executes `reporting.CreateReporters()` first?

If `ServiceControlRawDataReporting.OnStart()` is called first:

* `reportingOptions.OnCreateReporters()` is called, passing the lambda that creates them
  * `createReportersCalled` will be false, triggering the `else` block, storing the lambda.
* `reporting.CreateReporters()` is called
  * `createReportersCalled` set to true
  * `the lambda is stored, so it is executed and then set to null

If `reporting.CreateReporters()` is called first:

* `reporting.CreateReporters()`
  * `createReportersCalled` set to true
  *  `createMetricReporters` lambda is null, nothing else happens there
*  `ServiceControlRawDataReporting.OnStart()` is called by the feature, passing in the lambda
  * `createReportersCalled` is true, so the lambda is immediately executed.

So I don't think there are any holes. My only concern is if there could potentially be a race condition that would be worth wrapping with a thread lock, but given that this is initialization code I don't think that's possible.